Fix Ignore External Changes Bug

jorgekorgut opened the pull request:
Fix Issue #106410 : Add a check for external modified scene files (_is_scene_externally_modified()) before handling the logic of ignoring external changes.


![ignore-external-changes-test](https://github.com/user-attachments/assets/195c56d3-a764-41c3-a06c-dd0c28cd5c36)

* *Bugsquad edit, fixes: #106410*

KoBeWi commented:
As I suggested in the issue, it would be better to save only the scenes in the list, instead of using `_save_all_scenes()`.

jorgekorgut commented:
@KoBeWi Thanks for your feedback.

I have been reading the code for a while and I found just one reference containing the information of external unsaved files. This tree, called```disk_changed_list```, keeps reference of the string of the scenes names ? (not so sure about it) and I think it is only used for the UI.

Just to be sure you have spotted, calling the ```_save_all_scenes()``` does actually the saving scene logic only for the lists that are not saved in the editor ```_is_scene_unsaved()``` AND with this pull request scenes that were externally modified ```_is_scene_externally_modified()```.

Do you still would prefer a new function like ```_save_scene(i)``` that is called somewhere ? (dont know where or how do you want this function to be called)

KoBeWi commented:
`_save_scene()` already exists and takes a scene path (index is optional), so you can use that.

The reason I think it's better to save only modified scenes is that this information is already available. `_save_all_scenes()` is used in other places too (e.g. when running project), and with your current code it will perform unnecessary check.

I'd go with your suggestion 1. You can add a `LocalVector<String>` with paths of the scenes (the same ones as added to `disk_changed_list`).
Alternatively you can reuse `disk_changed_list`, since it already has all the scene paths. However it may contain not only scenes, so you'd have to mark items whether they are a scene or not. While it would work, the above solution sounds cleaner.

jorgekorgut committed:
Fix Ignore External Changes Bug

Fix Issue #106410 : Add a new list `disk_changed_scenes()` and a boolean `disk_changed_project` to keep track of modfied scenes and the project. Save them only if they are in the list or the boolean value is true, when ignoring external changes.

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -1304,21 +1304,28 @@ void EditorNode::_scan_external_changes() {
-			ti->set_text(0, editor_data.get_scene_path(i).get_file());
+			ti->set_text(0, scene_path.get_file());
 			need_reload = true;
+
```
```suggestion
```
Unnecessary blank line. Same below.

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -1304,21 +1304,28 @@ void EditorNode::_scan_external_changes() {
 		Ref<DirAccess> da = DirAccess::create(DirAccess::ACCESS_RESOURCES);
-		if (editor_data.get_scene_path(i) == "" || !da->file_exists(editor_data.get_scene_path(i))) {
+
+		String scene_path = editor_data.get_scene_path(i);
+
```
```suggestion
		const String scene_path = editor_data.get_scene_path(i);
```

Repiteo commented:
Thanks!
